CDCgov/prime-reportstream
# 8434 Extract LIVD Lookup FHIR custom function outside the conversion library

This PR extracts the LIVD lookup FHIR custom function outside the conversion library. It also adds the ability to pass an additional set of custom FHIR functions to the conversion library. The LIVD lookup table needed to be extracted because it makes calls to the ReportStream database and can't be part of the conversion library once it's moved to its own repo.

**If you are suggesting a fix for a currently exploitable issue, please disclose the issue to the prime-reportstream team directly outside of GitHub instead of filing a PR, so we may immediately patch the affected systems before a disclosure. See [SECURITY.md/Reporting a Vulnerability](https://github.com/CDCgov/prime-reportstream/blob/master/SECURITY.md#reporting-a-vulnerability) for more information.**

Test Steps:
1. The LIVD lookup function isn't being used in our mappings currently. The best way to test this is by running unit tests and checking for code coverage.
2. The test  `test convert with context` in `FhirToHl7ConverterTests.kt` shows how this function can be used in the FHIR to HL7 transforms and shows how we can pass additional FHIR custom functions to the `FhirToHl7Converter`
3. Test the LIVD lookup function using the Prime CLI.
4. Make sure your DB environment variables point to your local postgres DB
    `export POSTGRES_URL=jdbc:postgresql://localhost:5432/prime_data_hub`
    `export POSTGRES_USER=prime`
     `export POSTGRES_PASSWORD=changeIT!`
5. Start the Prime CLI with this command: `./prime fhirpath --input-file src/testIntegration/resources/datatests/FHIR_to_HL7/sample_SR_1_20230302-0001.fhir`
6. Execute this FHIR path command: `Bundle.entry.resource.ofType(Observation)[0].livdTableLookup("Test Performed LOINC Code")
` and make sure it returns results.

## Changes
- Extracts the LIVD Lookup FHIR custom function outside the conversion library
- Adds the ability to additional FHIR custom functions to the FHIR to HL7 conversion library
- Added unit tests

## Checklist

### Testing

### Process

## Linked Issues
- #8434 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
